File: src/jdwp/commands.ts
// Typed wrappers over the JDWP commands cappu's debugger issues. Each builds the
// command body with a ByteWriter (ID widths from the negotiated IdSizes) and
// parses the reply with a ByteReader. Locations and values have their own
// codecs because several commands embed them.
//
// Port reference for togo/internal/jdwp/commands.go.

import { ByteReader, ByteWriter, type IdSizes } from "./idCodec.ts";
import type { JdwpClient } from "./jdwpClient.ts";
import {
  CommandSet,
  EventRequestCmd,
  MethodCmd,
  ObjectReferenceCmd,
  ReferenceTypeCmd,
  StackFrameCmd,
  StringReferenceCmd,
  Tag,
  ThreadReferenceCmd,
  VirtualMachineCmd,
} from "./protocol.ts";

export interface Location {
  typeTag: number;
  classId: bigint;
  methodId: bigint;
  index: bigint;
}

export interface Frame {
  frameId: bigint;
  location: Location;
}

export interface MethodInfo {
  methodId: bigint;
  name: string;
  signature: string;
  modBits: number;
}

export interface LineTableEntry {
  lineCodeIndex: bigint;
  lineNumber: number;
}

export interface LineTable {
  start: bigint;
  end: bigint;
  lines: LineTableEntry[];
}

export interface ClassInfo {
  refTypeTag: number;
  typeId: bigint;
  status: number;
}

/** A decoded JDWP tagged value: an object reference (expandable) or a leaf. */
export type JdwpValue =
  | { tag: number; kind: "object"; objectId: bigint }
  | { tag: number; kind: "primitive"; value: number | bigint | boolean };

export function writeLocation(w: ByteWriter, sizes: IdSizes, loc: Location): void {
  w.u1(loc.typeTag)
    .id(loc.classId, sizes.referenceTypeID)
    .id(loc.methodId, sizes.methodID)
    .u8(loc.index);
}

export function readLocation(r: ByteReader, sizes: IdSizes): Location {
  return {
    typeTag: r.u1(),
    classId: r.id(sizes.referenceTypeID),
    methodId: r.id(sizes.methodID),
    index: r.u8(),
  };
}

const OBJECT_TAGS = new Set<number>([Tag.ARRAY, Tag.OBJECT, Tag.STRING, Tag.THREAD]);

/** Read a JDWP tagged value: leading tag byte then the width that tag implies. */
export function readValue(r: ByteReader, sizes: IdSizes): JdwpValue {
  const tag = r.u1();
  if (OBJECT_TAGS.has(tag)) return { tag, kind: "object", objectId: r.id(sizes.objectID) };
  switch (tag) {
    case Tag.BOOLEAN:
      return { tag, kind: "primitive", value: r.boolean() };
    case Tag.BYTE:
      return { tag, kind: "primitive", value: (r.u1() << 24) >> 24 };
    case Tag.CHAR:
      return { tag, kind: "primitive", value: r.u2() };
    case Tag.SHORT:
      return { tag, kind: "primitive", value: (r.u2() << 16) >> 16 };
    case Tag.INT:
      return { tag, kind: "primitive", value: r.i4() };
    case Tag.FLOAT:
      return { tag, kind: "primitive", value: r.bytes(4).readFloatBE(0) };
    case Tag.LONG:
      return { tag, kind: "primitive", value: BigInt.asIntN(64, r.u8()) };
    case Tag.DOUBLE:
      return { tag, kind: "primitive", value: r.bytes(8).readDoubleBE(0) };
    case Tag.VOID:
      return { tag, kind: "primitive", value: 0 };
    default:
      return { tag, kind: "object", objectId: r.id(sizes.objectID) };
  }
}

// --- VirtualMachine ---------------------------------------------------------

export function vmResume(c: JdwpClient): Promise<Buffer> {
  return c.send(CommandSet.VirtualMachine, VirtualMachineCmd.Resume);
}
export function vmSuspend(c: JdwpClient): Promise<Buffer> {
  return c.send(CommandSet.VirtualMachine, VirtualMachineCmd.Suspend);
}
export function vmExit(c: JdwpClient, code: number): Promise<Buffer> {
  return c.send(
    CommandSet.VirtualMachine,
    VirtualMachineCmd.Exit,
    new ByteWriter().i4(code).toBuffer(),
  );
}
export function vmDispose(c: JdwpClient): Promise<Buffer> {
  return c.send(CommandSet.VirtualMachine, VirtualMachineCmd.Dispose);
}

export async function allThreads(c: JdwpClient): Promise<bigint[]> {
  const r = new ByteReader(await c.send(CommandSet.VirtualMachine, VirtualMachineCmd.AllThreads));
  const n = r.u4();
  const ids: bigint[] = [];
  for (let i = 0; i < n; i++) ids.push(r.id(c.idSizes.objectID));
  return ids;
}

export async function classesBySignature(c: JdwpClient, signature: string): Promise<ClassInfo[]> {
  const body = new ByteWriter().string(signature).toBuffer();
  const r = new ByteReader(
    await c.send(CommandSet.VirtualMachine, VirtualMachineCmd.ClassesBySignature, body),
  );
  const n = r.u4();
  const out: ClassInfo[] = [];
  for (let i = 0; i < n; i++) {
    out.push({ refTypeTag: r.u1(), typeId: r.id(c.idSizes.referenceTypeID), status: r.i4() });
  }
  return out;
}

// --- ThreadReference --------------------------------------------------------

export async function threadName(c: JdwpClient, threadId: bigint): Promise<string> {
  const body = new ByteWriter().id(threadId, c.idSizes.objectID).toBuffer();
  return new ByteReader(
    await c.send(CommandSet.ThreadReference, ThreadReferenceCmd.Name, body),
  ).string();
}

export async function threadFrames(
  c: JdwpClient,
  threadId: bigint,
  start = 0,
  length = -1,
): Promise<Frame[]> {
  const body = new ByteWriter().id(threadId, c.idSizes.objectID).i4(start).i4(length).toBuffer();
  const r = new ByteReader(
    await c.send(CommandSet.ThreadReference, ThreadReferenceCmd.Frames, body),
  );
  const n = r.u4();
  const frames: Frame[] = [];
  for (let i = 0; i < n; i++) {
    frames.push({ frameId: r.id(c.idSizes.frameID), location: readLocation(r, c.idSizes) });
  }
  return frames;
}

export function threadResume(c: JdwpClient, threadId: bigint): Promise<Buffer> {
  const body = new ByteWriter().id(threadId, c.idSizes.objectID).toBuffer();
  return c.send(CommandSet.ThreadReference, ThreadReferenceCmd.Resume, body);
}

// --- ReferenceType / Method -------------------------------------------------

export async function referenceTypeSignature(c: JdwpClient, classId: bigint): Promise<string> {
  const body = new ByteWriter().id(classId, c.idSizes.referenceTypeID).toBuffer();
  return new ByteReader(
    await c.send(CommandSet.ReferenceType, ReferenceTypeCmd.Signature, body),
  ).string();
}

export async function referenceTypeMethods(c: JdwpClient, classId: bigint): Promise<MethodInfo[]> {
  const body = new ByteWriter().id(classId, c.idSizes.referenceTypeID).toBuffer();
  const r = new ByteReader(await c.send(CommandSet.ReferenceType, ReferenceTypeCmd.Methods, body));
  const n = r.u4();
  const methods: MethodInfo[] = [];
  for (let i = 0; i < n; i++) {
    methods.push({
      methodId: r.id(c.idSizes.methodID),
      name: r.string(),
      signature: r.string(),
      modBits: r.i4(),
    });
  }
  return methods;
}

export interface VariableSlot {
  codeIndex: bigint;
  name: string;
  signature: string;
  length: number;
  slot: number;
}

export async function methodVariableTable(
  c: JdwpClient,
  classId: bigint,
  methodId: bigint,
): Promise<VariableSlot[]> {
  const body = new ByteWriter()
    .id(classId, c.idSizes.referenceTypeID)
    .id(methodId, c.idSizes.methodID)
    .toBuffer();
  const r = new ByteReader(await c.send(CommandSet.Method, MethodCmd.VariableTable, body));
  r.u4(); // argCnt (unused: we filter by code-index visibility, not arg count)
  const n = r.u4();
  const slots: VariableSlot[] = [];
  for (let i = 0; i < n; i++) {
    slots.push({
      codeIndex: r.u8(),
      name: r.string(),
      signature: r.string(),
      length: r.i4(),
      slot: r.i4(),
    });
  }
  return slots;
}

export async function methodLineTable(
  c: JdwpClient,
  classId: bigint,
  methodId: bigint,
): Promise<LineTable> {
  const body = new ByteWriter()
    .id(classId, c.idSizes.referenceTypeID)
    .id(methodId, c.idSizes.methodID)
    .toBuffer();
  const r = new ByteReader(await c.send(CommandSet.Method, MethodCmd.LineTable, body));
  const start = r.u8();
  const end = r.u8();
  const n = r.u4();
  const lines: LineTableEntry[] = [];
  for (let i = 0; i < n; i++) lines.push({ lineCodeIndex: r.u8(), lineNumber: r.i4() });
  return { start, end, lines };
}

// --- EventRequest -----------------------------------------------------------

/**
 * One EventRequest.Set modifier; the wire payload depends on its kind. The
 * literal kinds match ModifierKind in protocol.ts (Count=1, ClassMatch=5,
 * LocationOnly=7, Step=10).
 */
export type Modifier =
  | { kind: 1; count: number }
  | { kind: 5; pattern: string }
  | { kind: 7; location: Location }
  | { kind: 10; threadId: bigint; size: number; depth: number };

function writeModifier(w: ByteWriter, sizes: IdSizes, m: Modifier): void {
  w.u1(m.kind);
  switch (m.kind) {
    case 1:
      w.i4(m.count);
      break;
    case 5:
      w.string(m.pattern);
      break;
    case 7:
      writeLocation(w, sizes, m.location);
      break;
    case 10:
      w.id(m.threadId, sizes.objectID).i4(m.size).i4(m.depth);
      break;
  }
}

export async function eventRequestSet(
  c: JdwpClient,
  eventKind: number,
  suspendPolicy: number,
  modifiers: Modifier[],
): Promise<number> {
  const w = new ByteWriter().u1(eventKind).u1(suspendPolicy).i4(modifiers.length);
  for (const m of modifiers) writeModifier(w, c.idSizes, m);
  const r = new ByteReader(
    await c.send(CommandSet.EventRequest, EventRequestCmd.Set, w.toBuffer()),
  );
  return r.i4();
}

export function eventRequestClear(
  c: JdwpClient,
  eventKind: number,
  requestId: number,
): Promise<Buffer> {
  const body = new ByteWriter().u1(eventKind).i4(requestId).toBuffer();
  return c.send(CommandSet.EventRequest, EventRequestCmd.Clear, body);
}

// --- StackFrame / values ----------------------------------------------------

export async function stackFrameGetValues(
  c: JdwpClient,
  threadId: bigint,
  frameId: bigint,
  slots: { slot: number; sigByte: number }[],
): Promise<JdwpValue[]> {
  const w = new ByteWriter()
    .id(threadId, c.idSizes.objectID)
    .id(frameId, c.idSizes.frameID)
    .i4(slots.length);
  for (const s of slots) w.i4(s.slot).u1(s.sigByte);
  const r = new ByteReader(
    await c.send(CommandSet.StackFrame, StackFrameCmd.GetValues, w.toBuffer()),
  );
  const n = r.u4();
  const values: JdwpValue[] = [];
  for (let i = 0; i < n; i++) values.push(readValue(r, c.idSizes));
  return values;
}

export async function stringValue(c: JdwpClient, stringId: bigint): Promise<string> {
  const body = new ByteWriter().id(stringId, c.idSizes.objectID).toBuffer();
  return new ByteReader(
    await c.send(CommandSet.StringReference, StringReferenceCmd.Value, body),
  ).string();
}

export async function objectReferenceType(
  c: JdwpClient,
  objectId: bigint,
): Promise<{ refTypeTag: number; typeId: bigint }> {
  const body = new ByteWriter().id(objectId, c.idSizes.objectID).toBuffer();
  const r = new ByteReader(
    await c.send(CommandSet.ObjectReference, ObjectReferenceCmd.ReferenceType, body),
  );
  return { refTypeTag: r.u1(), typeId: r.id(c.idSizes.referenceTypeID) };
}

File: togo/internal/jdwp/commands.go
package jdwp

// Typed wrappers over the JDWP commands cappu's debugger issues. Each builds the
// command body with a Writer (ID widths from the negotiated IDSizes) and parses
// the reply with a Reader. Locations and values have their own codecs because
// several commands embed them. Port of src/jdwp/commands.ts.

import "math"

type Location struct {
	TypeTag  byte
	ClassID  uint64
	MethodID uint64
	Index    uint64
}

type Frame struct {
	FrameID  uint64
	Location Location
}

type MethodInfo struct {
	MethodID  uint64
	Name      string
	Signature string
	ModBits   int32
}

type LineTableEntry struct {
	LineCodeIndex uint64
	LineNumber    int32
}

type LineTable struct {
	Start uint64
	End   uint64
	Lines []LineTableEntry
}

type ClassInfo struct {
	RefTypeTag byte
	TypeID     uint64
	Status     int32
}

type VariableSlot struct {
	CodeIndex uint64
	Name      string
	Signature string
	Length    int32
	Slot      int32
}

// Value is a decoded JDWP tagged value. Object is true for reference types
// (the ObjectID is then meaningful); otherwise the primitive payload is in the
// field matching Tag.
type Value struct {
	Tag      byte
	Object   bool
	ObjectID uint64
	Int      int64
	Float    float64
	Bool     bool
}

func writeLocation(w *Writer, s IDSizes, loc Location) {
	w.U1(loc.TypeTag).ID(loc.ClassID, s.ReferenceTypeID).ID(loc.MethodID, s.MethodID).U8(loc.Index)
}

func readLocation(r *Reader, s IDSizes) Location {
	return Location{
		TypeTag:  r.U1(),
		ClassID:  r.ID(s.ReferenceTypeID),
		MethodID: r.ID(s.MethodID),
		Index:    r.U8(),
	}
}

func isObjectTag(tag byte) bool {
	switch tag {
	case TagArray, TagObject, TagString, TagThread:
		return true
	}
	return false
}

// ReadValue reads a JDWP tagged value: leading tag byte then the implied width.
func ReadValue(r *Reader, s IDSizes) Value {
	tag := r.U1()
	if isObjectTag(tag) {
		return Value{Tag: tag, Object: true, ObjectID: r.ID(s.ObjectID)}
	}
	switch tag {
	case TagBoolean:
		return Value{Tag: tag, Bool: r.Bool()}
	case TagByte:
		return Value{Tag: tag, Int: int64(int8(r.U1()))}
	case TagChar:
		return Value{Tag: tag, Int: int64(r.U2())}
	case TagShort:
		return Value{Tag: tag, Int: int64(int16(r.U2()))}
	case TagInt:
		return Value{Tag: tag, Int: int64(r.I4())}
	case TagLong:
		return Value{Tag: tag, Int: int64(r.U8())}
	case TagFloat:
		return Value{Tag: tag, Float: float64(math.Float32frombits(r.U4()))}
	case TagDouble:
		return Value{Tag: tag, Float: math.Float64frombits(r.U8())}
	case TagVoid:
		return Value{Tag: tag}
	default:
		return Value{Tag: tag, Object: true, ObjectID: r.ID(s.ObjectID)}
	}
}

// --- VirtualMachine ---------------------------------------------------------

func VMResumeCmd(c *Client) error  { _, err := c.Send(CSVirtualMachine, VMResume, nil); return err }
func VMSuspendCmd(c *Client) error { _, err := c.Send(CSVirtualMachine, VMSuspend, nil); return err }
func VMDisposeCmd(c *Client) error { _, err := c.Send(CSVirtualMachine, VMDispose, nil); return err }

func VMExitCmd(c *Client, code int32) error {
	w := &Writer{}
	w.I4(code)
	_, err := c.Send(CSVirtualMachine, VMExit, w.Buffer())
	return err
}

func AllThreads(c *Client) ([]uint64, error) {
	data, err := c.Send(CSVirtualMachine, VMAllThreads, nil)
	if err != nil {
		return nil, err
	}
	r := NewReader(data)
	n := int(r.U4())
	ids := make([]uint64, n)
	for i := range n {
		ids[i] = r.ID(c.IDSizes.ObjectID)
	}
	return ids, nil
}

func ClassesBySignature(c *Client, signature string) ([]ClassInfo, error) {
	w := &Writer{}
	w.String(signature)
	data, err := c.Send(CSVirtualMachine, VMClassesBySignature, w.Buffer())
	if err != nil {
		return nil, err
	}
	r := NewReader(data)
	n := int(r.U4())
	out := make([]ClassInfo, n)
	for i := range n {
		out[i] = ClassInfo{RefTypeTag: r.U1(), TypeID: r.ID(c.IDSizes.ReferenceTypeID), Status: r.I4()}
	}
	return out, nil
}

// --- ThreadReference --------------------------------------------------------

func ThreadName(c *Client, threadID uint64) (string, error) {
	w := &Writer{}
	w.ID(threadID, c.IDSizes.ObjectID)
	data, err := c.Send(CSThreadReference, TRName, w.Buffer())
	if err != nil {
		return "", err
	}
	return NewReader(data).String(), nil
}

func ThreadFrames(c *Client, threadID uint64, start, length int32) ([]Frame, error) {
	w := &Writer{}
	w.ID(threadID, c.IDSizes.ObjectID).I4(start).I4(length)
	data, err := c.Send(CSThreadReference, TRFrames, w.Buffer())
	if err != nil {
		return nil, err
	}
	r := NewReader(data)
	n := int(r.U4())
	frames := make([]Frame, n)
	for i := range n {
		frames[i] = Frame{FrameID: r.ID(c.IDSizes.FrameID), Location: readLocation(r, c.IDSizes)}
	}
	return frames, nil
}

// --- ReferenceType / Method -------------------------------------------------

func ReferenceTypeSignature(c *Client, classID uint64) (string, error) {
	w := &Writer{}
	w.ID(classID, c.IDSizes.ReferenceTypeID)
	data, err := c.Send(CSReferenceType, RTSignature, w.Buffer())
	if err != nil {
		return "", err
	}
	return NewReader(data).String(), nil
}

func ReferenceTypeMethods(c *Client, classID uint64) ([]MethodInfo, error) {
	w := &Writer{}
	w.ID(classID, c.IDSizes.ReferenceTypeID)
	data, err := c.Send(CSReferenceType, RTMethods, w.Buffer())
	if err != nil {
		return nil, err
	}
	r := NewReader(data)
	n := int(r.U4())
	methods := make([]MethodInfo, n)
	for i := range n {
		methods[i] = MethodInfo{
			MethodID:  r.ID(c.IDSizes.MethodID),
			Name:      r.String(),
			Signature: r.String(),
			ModBits:   r.I4(),
		}
	}
	return methods, nil
}

func MethodVariableTable(c *Client, classID, methodID uint64) ([]VariableSlot, error) {
	w := &Writer{}
	w.ID(classID, c.IDSizes.ReferenceTypeID).ID(methodID, c.IDSizes.MethodID)
	data, err := c.Send(CSMethod, MVariableTable, w.Buffer())
	if err != nil {
		return nil, err
	}
	r := NewReader(data)
	r.U4() // argCnt (unused: visibility is by code-index range, not arg count)
	n := int(r.U4())
	slots := make([]VariableSlot, n)
	for i := range n {
		slots[i] = VariableSlot{
			CodeIndex: r.U8(),
			Name:      r.String(),
			Signature: r.String(),
			Length:    r.I4(),
			Slot:      r.I4(),
		}
	}
	return slots, nil
}

func MethodLineTableCmd(c *Client, classID, methodID uint64) (LineTable, error) {
	w := &Writer{}
	w.ID(classID, c.IDSizes.ReferenceTypeID).ID(methodID, c.IDSizes.MethodID)
	data, err := c.Send(CSMethod, MLineTable, w.Buffer())
	if err != nil {
		return LineTable{}, err
	}
	r := NewReader(data)
	lt := LineTable{Start: r.U8(), End: r.U8()}
	n := int(r.U4())
	lt.Lines = make([]LineTableEntry, n)
	for i := range n {
		lt.Lines[i] = LineTableEntry{LineCodeIndex: r.U8(), LineNumber: r.I4()}
	}
	return lt, nil
}

// --- EventRequest -----------------------------------------------------------

// Modifier is one EventRequest.Set modifier; the wire payload depends on Kind
// (Count, ClassMatch, LocationOnly or Step).
type Modifier struct {
	Kind      byte
	Count     int32
	Pattern   string
	Location  Location
	ThreadID  uint64
	StepSize  int32
	StepDepth int32
}

func writeModifier(w *Writer, s IDSizes, m Modifier) {
	w.U1(m.Kind)
	switch m.Kind {
	case ModCount:
		w.I4(m.Count)
	case ModClassMatch:
		w.String(m.Pattern)
	case ModLocationOnly:
		writeLocation(w, s, m.Location)
	case ModStep:
		w.ID(m.ThreadID, s.ObjectID).I4(m.StepSize).I4(m.StepDepth)
	}
}

func EventRequestSet(c *Client, eventKind, suspendPolicy byte, modifiers []Modifier) (int32, error) {
	w := &Writer{}
	w.U1(eventKind).U1(suspendPolicy).I4(int32(len(modifiers)))
	for _, m := range modifiers {
		writeModifier(w, c.IDSizes, m)
	}
	data, err := c.Send(CSEventRequest, ERSet, w.Buffer())
	if err != nil {
		return 0, err
	}
	return NewReader(data).I4(), nil
}

func EventRequestClear(c *Client, eventKind byte, requestID int32) error {
	w := &Writer{}
	w.U1(eventKind).I4(requestID)
	_, err := c.Send(CSEventRequest, ERClear, w.Buffer())
	return err
}

// --- StackFrame / values ----------------------------------------------------

type Slot struct {
	Slot    int32
	SigByte byte
}

func StackFrameGetValues(c *Client, threadID, frameID uint64, slots []Slot) ([]Value, error) {
	w := &Writer{}
	w.ID(threadID, c.IDSizes.ObjectID).ID(frameID, c.IDSizes.FrameID).I4(int32(len(slots)))
	for _, s := range slots {
		w.I4(s.Slot).U1(s.SigByte)
	}
	data, err := c.Send(CSStackFrame, SFGetValues, w.Buffer())
	if err != nil {
		return nil, err
	}
	r := NewReader(data)
	n := int(r.U4())
	values := make([]Value, n)
	for i := range n {
		values[i] = ReadValue(r, c.IDSizes)
	}
	return values, nil
}

func StringValue(c *Client, stringID uint64) (string, error) {
	w := &Writer{}
	w.ID(stringID, c.IDSizes.ObjectID)
	data, err := c.Send(CSStringReference, SRValue, w.Buffer())
	if err != nil {
		return "", err
	}
	return NewReader(data).String(), nil
}
